crypto/compactcert/builder.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+//msgp:ignore sigslot
+type sigslot struct {
+	// part is the participant signing this message.  The participant
+	// information is tracked here for convenience, but it does not
+	// appear in the commitment to the sigs array.
+	part Participant
+
+	// Include the parts of the sigslot that form the commitment to
+	// the sigs array.
+	sigslotCommit
+}
+
+// Builder keeps track of signatures on a message and eventually produces
+// a compact certificate for that message.
+type Builder struct {
+	Params
+
+	sigs         []sigslot // Indexed by pos in participants
+	signedWeight uint64    // Total weight of signatures so far
+	participants []Participant
+	parttree     *merklearray.Tree
+
+	// Cached cert, if Build() was called and no subsequent
+	// Add() calls were made.
+	cert *Cert
+}
+
+// MkBuilder constructs an empty builder (with no signatures).  The message
+// to be signed, as well as other security parameters, are specified in
+// param.  The participants that will sign the message are in part and
+// parttree.
+func MkBuilder(param Params, part []Participant, parttree *merklearray.Tree) (*Builder, error) {
+	npart := len(part)
+
+	b := &Builder{
+		Params:       param,
+		sigs:         make([]sigslot, npart),
+		signedWeight: 0,
+		participants: part,
+		parttree:     parttree,
+	}
+
+	return b, nil
+}
+
+// Present checks if the builder already contains a signature at a particular
+// offset.
+func (b *Builder) Present(pos uint64) bool {
+	return b.sigs[pos].part.Weight != 0
+}
+
+// Add a signature to the set of signatures available for building a certificate.
+// verifySig should be set to true in production; setting it to false is useful
+// for benchmarking to avoid the cost of signature checks.
+func (b *Builder) Add(pos uint64, sig crypto.OneTimeSignature, verifySig bool) error {
+	if b.Present(pos) {
+		return fmt.Errorf("position %d already added", pos)
+	}
+
+	// Check participants array
+	if pos >= uint64(len(b.participants)) {
+		return fmt.Errorf("pos %d >= len(participants) %d", pos, len(b.participants))
+	}
+
+	p := b.participants[pos]
+
+	if p.Weight == 0 {
+		return fmt.Errorf("position %d has zero weight", pos)
+	}
+
+	// Check signature
+	ephID := basics.OneTimeIDForRound(b.SigRound, p.KeyDilution)
+	if verifySig && !p.PK.Verify(ephID, b.Msg, sig) {
+		return fmt.Errorf("signature does not verify under ID %v", ephID)
+	}
+
+	// Remember the signature
+	b.sigs[pos].part = p
+	b.sigs[pos].Sig.OneTimeSignature = sig
+	b.signedWeight += p.Weight
+	b.cert = nil
+	return nil
+}
+
+// Ready returns whether the certificate is ready to be built.
+func (b *Builder) Ready() bool {
+	return b.signedWeight >= b.Params.ProvenWeight
+}
+
+// SignedWeight returns the total weight of signatures added so far.
+func (b *Builder) SignedWeight() uint64 {
+	return b.signedWeight
+}
+
+//msgp:ignore sigCommit
+type sigCommit []sigslot
```

Comment:
I find this name confusing -- sigCommit seems like a signature commitment

Reply:
What about `sigsArray` or `sigCommitArray`?

Reply:
sigsToCommit?

Reply:
That seems clear; thanks.